common/src/main/java/com/wynntils/models/abilitytree/AbilityTreeContainerQueries.java

```diff
@@ -0,0 +1,174 @@
+                .matchTitle(Models.Container.ABILITY_TREE_PATTERN.pattern())
+                .processContainer(c -> {});
+
+        // region Hack for going back to first page without knowing our current page
+
+        for (int i = Models.AbilityTree.ABILITY_TREE_PAGES - 1; i > 0; i--) {
+            queryBuilder
+                    .clickOnSlotIfExists(PREVIOUS_PAGE_SLOT, DUMMY_SLOT)
+                    .matchTitle(Models.Container.ABILITY_TREE_PATTERN.pattern())
+                    .processContainer(c -> {});
+        }
+
+        // endregion
+
+        // We are on the first page now
+        queryBuilder
+                .clickOnSlot(DUMMY_SLOT)
+                .matchTitle(Models.Container.ABILITY_TREE_PATTERN.pattern())
+                .processContainer(c -> processor.processPage(c, 1));
+
+        // region Hack for setWaitForMenuReopen to work with page 2
+        // Going from first page to second adds a new button, "previous". This triggers processContainer prematurely.
+
+        queryBuilder
+                .clickOnSlotWithName(NEXT_PAGE_SLOT, Items.STONE_AXE, NEXT_PAGE_ITEM_NAME)
+                .matchTitle(Models.Container.ABILITY_TREE_PATTERN.pattern())
+                .processContainer(c -> {});
+        queryBuilder
+                .clickOnSlot(DUMMY_SLOT)
+                .matchTitle(Models.Container.ABILITY_TREE_PATTERN.pattern())
+                .processContainer(c -> processor.processPage(c, 2));
+
+        // endregion
+
+        for (int i = 3; i <= Models.AbilityTree.ABILITY_TREE_PAGES; i++) {
+            final int page = i; // Lambdas need final variables
+            queryBuilder
+                    .clickOnSlotWithName(NEXT_PAGE_SLOT, Items.STONE_AXE, NEXT_PAGE_ITEM_NAME)
+                    .matchTitle(Models.Container.ABILITY_TREE_PATTERN.pattern())
+                    .processContainer(c -> processor.processPage(c, page));
+        }
+
+        queryBuilder.build().executeQuery();
+    }
+
+    @FunctionalInterface
+    public interface AbilityTreeProcessor {
+        void processPage(ContainerContent content, int page);
+    }
+
+    /**
+     * Parses the whole ability tree and saves it to disk.
+     */
+    public static class AbilityPageDumper implements AbilityTreeProcessor {
+        private final Consumer<AbilityTreeInfo> supplier;
+        private final UnprocessedAbilityTreeInfo unprocessedTree = new UnprocessedAbilityTreeInfo();
+
+        public AbilityPageDumper(Consumer<AbilityTreeInfo> supplier) {
+            this.supplier = supplier;
+        }
+
+        @Override
+        public void processPage(ContainerContent content, int page) {
+            List<ItemStack> items = content.items();
+
+            for (int slot = 0; slot < items.size(); slot++) {
+                ItemStack itemStack = items.get(slot);
+
+                unprocessedTree.processItem(itemStack, page, slot, true);
+            }
+
+            boolean lastPage = page == Models.AbilityTree.ABILITY_TREE_PAGES;
```

Comment:
Using both = and == in the same line is always a bit confusing. In this case, the variable lastPage is strictly not needed, and I think the comparison is better inlined in the if test. If you want to explain that you are looking for the last page, maybe to so in a short comment instead?

Reply:
Yes I can refactor this back, this boolean used to be used for many stuff before.